Bio/SearchIO/InfernalIO/infernal_text.py

```diff
@@ -0,0 +1,543 @@
+# Copyright 2024 by Samuel Prince. All rights reserved.
+#
+# This file is part of the Biopython distribution and governed by your
+# choice of the "Biopython License Agreement" or the "BSD 3-Clause License".
+# Please see the LICENSE file that should have been included as part of this
+# package.
+
+
+"""Bio.SearchIO parser for Infernal plain text output format."""
+
+import re
+import operator
+
+from Bio.SearchIO._index import SearchIndexer
+from Bio.SearchIO._model import HSP
+from Bio.SearchIO._model import HSPFragment
+from Bio.SearchIO._model import QueryResult
+from Bio.SearchIO._utils import read_forward
+
+from ._base import _BaseInfernalParser
+
+__all__ = ("InfernalTextParser", "InfernalTextIndexer")
+
+# precompile regex patterns for faster processing
+# program name
+_RE_PROGRAM = re.compile(r"^# .*?(\w?cm\w+) :: .*$")
+# version string
+_RE_VERSION = re.compile(r"# INFERNAL+ ([\w+\.]{2,}) .*$")
+# option string
+_RE_OPT = re.compile(r"^# (.+):\s+(.+)$")
+# numbers only
+_RE_NUMERIC = re.compile(r"\d+")
+_RE_NOT_NUMERIC = re.compile(r"\D")
+# letters only
+_RE_LETTERS = re.compile(r"[^A-Za-z]")
+# missing segments in the alignment block
+_RE_SPLIT_ALN = re.compile(r"(\*\[[ 0-9]+\]\*)+")
+
+# divider for Infernal plain text output, these divider are
+# always taken from the beginning of the line to be usable in
+# self.line.startswith() method
+# header options
+_DIV_HEADER_OPT = "# - - -"
+# query
+_DIV_QUERY_END = "//"
+_DIV_QUERY_START = "Query:"
+# hits
+_DIV_HITS_END_CM = "Internal CM pipeline statistics summary:"
+_DIV_HITS_END_HMM = "Internal HMM-only pipeline statistics summary:"
+_DIV_HIT_SCORE_TABLE = "Hit scores:"
+_DIV_HIT_ALIGNMENT = "Hit alignments:"
+_DIV_NO_HIT = "   [No hits detected that satisfy reporting thresholds]"
+_DIV_TABLE_START = " ----   --------- ------"
+_DIV_INC_THRESHOLD = " ------ inclusion threshold ------"
+# hit alignment
+_DIV_ALIGNMENT_START = ">> "
+
+
+class InfernalTextParser(_BaseInfernalParser):
+    """Parser for the Infernal text output."""
+
+    def __init__(self, handle):
+        """Initialize the class."""
+        self.handle = handle
+        self.line = read_forward(self.handle)
+        self._meta = self._parse_header()
+
+    def __iter__(self):
+        """Iterate over query results."""
+        yield from self._parse_qresult()
+
+    def _read_until(self, bool_func):
+        """Read the file handle until the given function returns True (PRIVATE)."""
+        while True:
+            if not self.line or bool_func(self.line):
+                return
+            else:
+                self.line = read_forward(self.handle)
```

Comment:
Also a style suggestion, since the `else` is redundant:

```suggestion
            if not self.line or bool_func(self.line):
                return
            self.line = read_forward(self.handle)
```